src/sql.rs

```diff
@@ -711,6 +711,44 @@ pub async fn housekeeping(context: &Context) -> Result<()> {
 
     context.schedule_quota_update().await?;
 
+    loop {
+        let mut have_uncompressed = false;
+        if let Err(e) = context
+            .sql
+            .transaction(|transaction| {
+                let mut update_stmt = transaction
+                    .prepare("UPDATE msgs SET mime_headers=?, mime_compressed=1 WHERE id=?")?;
+                let mut select_stmt = transaction.prepare(
+                    "\
+                    SELECT id, mime_headers FROM msgs WHERE mime_headers!='' AND mime_compressed=0 \
+                    LIMIT 50",
+                )?;
+                for row in select_stmt.query_map([], |row| {
+                    let id: crate::message::MsgId = row.get(0)?;
+                    let mime_headers: Vec<u8> = row.get(1)?;
+                    Ok((id, mime_headers))
+                })? {
+                    have_uncompressed = true;
+                    let (id, mime_headers) = row?;
+                    let mime_headers = buf_compress(&mime_headers)?;
```

Comment:
So, the problem is that we're blocking other write transactions for too long (blocking a db connection isn't a problem, we can just allocate one extra connection if so). Maybe it's ok to process one message per transaction if so? Otherwise it becomes complicated: select 50 messages, then compress them, then start a transaction to store them and also make sure there are no races (a message wasn't deleted and so on)

Reply:
I think better to introduce a semaphore with 2 permits around `Sql::write_mtx`. So, one permit is consumed by `Sql::write_lock()` and the second may be checked using `available_permits()` inside long-running transactions so that they can cancel running if a contention with other transactions is detected

Reply:
You can check it out now, it really looks simpler than splitting the transaction. The only downside is that it may be cancelled early and update just one row, but i don't think it will happen frequently.

Reply:
Increased the connection pool size to 4 and added a comment on this